# Demonstrating Different Decoding Strategies with Hugging Face Transformers
In this notebook, we will explore various decoding strategies for text generation using a small encoder-decoder model from Hugging Face's Transformers library. We'll apply these strategies to two different datasets:

- Translation task where deterministic strategies are expected to perform better.
- Summarization task where stochastic strategies might yield more diverse and informative outputs.

The decoding strategies we'll test include:

1. Greedy Search
2. Beam Search
3. Temperature Sampling
4. Top-k Sampling
5. Top-p (Nucleus) Sampling

We'll define a custom ```generate_text``` function to apply these strategies and evaluate their performance using appropriate metrics for each dataset.

Here are some useful links you might want to check:
- [Auto Classes](https://huggingface.co/docs/transformers/model_doc/auto)
- [transformers\AutoModel](https://huggingface.co/docs/transformers/model_doc/auto#transformers.AutoModel)
- [transformers\AutoTokenizer](https://huggingface.co/docs/transformers/model_doc/auto#transformers.AutoTokenizer)
- [Google's T5](https://arxiv.org/pdf/1910.10683)
- [T5-small](https://huggingface.co/google-t5/t5-small)
- [Flan-T5-small](https://huggingface.co/google/flan-t5-small)

In [1]:
# !pip install transformers datasets sacrebleu rouge_score evaluate --quiet


[notice] A new release of pip is available: 25.0.1 -> 25.2
[notice] To update, run: pip install --upgrade pip


In [2]:
import torch
from transformers import AutoModelForSeq2SeqLM, AutoTokenizer
from datasets import load_dataset
import random
import evaluate

/Users/mariacarreno/opt/anaconda3/lib/python3.9/site-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: 'dlopen(/Users/mariacarreno/opt/anaconda3/lib/python3.9/site-packages/torchvision/image.so, 0x0006): Symbol not found: __ZN3c1017RegisterOperatorsD1Ev
  Referenced from: <84D64A51-D060-3E0B-B3C3-176DC1865B41> /Users/mariacarreno/opt/anaconda3/lib/python3.9/site-packages/torchvision/image.so
  Expected in:     <709C1DF5-D253-3C66-87E2-C99FD3A259DF> /Users/mariacarreno/opt/anaconda3/lib/python3.9/site-packages/torch/lib/libtorch_cpu.dylib'If you don't plan on using image functionality from `torchvision.io`, you can ignore this warning. Otherwise, there might be something wrong with your environment. Did you have `libjpeg` or `libpng` installed before building `torchvision` from source?
  warn(


### Load the Pre-trained Model and Tokenizer

We'll use the ```flan-T5-small``` model, which is a small encoder-decoder model suitable for both story_gen and summarization tasks. This model is based on Google's ```t5-small``` model, but fine-tuned on more than 1000 additional tasks covering also more languages.

In [3]:
model_name = 'google/flan-t5-small'
model = AutoModelForSeq2SeqLM.from_pretrained(model_name)
tokenizer = AutoTokenizer.from_pretrained(model_name)


config.json: 0.00B [00:00, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model.safetensors:   0%|          | 0.00/308M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/147 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json: 0.00B [00:00, ?B/s]

In [4]:
def generate_text(input_text, strategy='greedy', max_length=50, **kwargs):
    """Generates text based on the specified decoding strategy.

    Args:
        input_text (str): The input text to be processed by the model.
        strategy (str, optional): The decoding strategy to use. Defaults to 'greedy'.
            Options include:
            - 'greedy': Greedy search decoding.
            - 'beam': Beam search decoding.
            - 'temperature': Temperature sampling.
            - 'top-k': Top-k sampling.
            - 'top-p': Top-p (nucleus) sampling.
            - 'contrastive': Contrastive search decoding.
        max_length (int, optional): The maximum length of the generated text. Defaults to 50.
        **kwargs: Additional keyword arguments specific to the decoding strategy.

    Keyword Args:
        num_beams (int, optional): Number of beams for beam search. Defaults to 5.
            Used when `strategy='beam'`.
        temperature (float, optional): Sampling temperature. Defaults to 1.0.
            Used when `strategy='temperature'`.
        top_k (int, optional): The number of highest probability vocabulary tokens to keep for top-k filtering. Defaults to 50.
            Used when `strategy='top-k'` or when `strategy='contrastive'`.
        top_p (float, optional): Cumulative probability for nucleus sampling. Defaults to 0.95.
            Used when `strategy='top-p'`.
        penalty_alpha (float, optional): Contrastive search penalty factor. Defaults to 0.6.
            Used when `strategy='contrastive'`.

    Returns:
        str: The generated text based on the decoding strategy.

    Raises:
        ValueError: If an unknown decoding strategy is specified.
    """
    input_ids = tokenizer.encode(input_text, return_tensors='pt')
    model.eval()
    with torch.no_grad():
        if strategy == 'greedy':
            output_ids = model.generate(input_ids, max_length=max_length)
        elif strategy == 'beam':
            num_beams = kwargs.get('num_beams', 5)
            output_ids = model.generate(
                input_ids, max_length=max_length, num_beams=num_beams, early_stopping=True
            )
        elif strategy == 'temperature':
            temperature = kwargs.get('temperature', 1.0)
            output_ids = model.generate(
                input_ids, max_length=max_length, do_sample=True, temperature=temperature
            )
        elif strategy == 'top-k':
            top_k = kwargs.get('top_k', 50)
            output_ids = model.generate(
                input_ids, max_length=max_length, do_sample=True, top_k=top_k
            )
        elif strategy == 'top-p':
            top_p = kwargs.get('top_p', 0.95)
            output_ids = model.generate(
                input_ids, max_length=max_length, do_sample=True, top_p=top_p
            )
        elif strategy == 'contrastive':
            penalty_alpha = kwargs.get('penalty_alpha', 0.6)
            top_k = kwargs.get('top_k', 4)
            output_ids = model.generate(
                input_ids, max_length=max_length, penalty_alpha=penalty_alpha, top_k=top_k
            )
        else:
            raise ValueError("Unknown strategy: {}".format(strategy))

    output_text = tokenizer.decode(output_ids[0], skip_special_tokens=True)
    return output_text


### Use case 1: Neural Machine Translation
The WMT16 English-German dataset is a collection of parallel sentences in English and German used for machine translation tasks. It is part of the Conference on Machine story_gen (WMT) shared tasks, which are benchmarks for evaluating machine story_gen systems. The dataset contains professionally translated sentences and covers a variety of topics, making it ideal for training and evaluating story_gen models.

We'll use a subset (1% of the test split) of the dataset for evaluation purposes.

In [5]:
dataset_nmt = load_dataset('wmt16', 'de-en', split='test[:1%]')

README.md: 0.00B [00:00, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


train-00000-of-00003.parquet:   0%|          | 0.00/282M [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


train-00001-of-00003.parquet:   0%|          | 0.00/267M [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


train-00002-of-00003.parquet:   0%|          | 0.00/277M [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


validation-00000-of-00001.parquet:   0%|          | 0.00/343k [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


test-00000-of-00001.parquet:   0%|          | 0.00/475k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/4548885 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/2169 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/2999 [00:00<?, ? examples/s]

In [9]:
# Randomly select a sample from the dataset
sample = random.choice(dataset_nmt['translation'])
input_text = sample['de']
target_text = sample['en']

print("Input Text:")
print(input_text)
print("\nTarget Text:")
print(target_text)


Input Text:
Die beiden wollten über die Umsetzung der internationalen Vereinbarung sowie über Teherans destabilisierende Maßnahmen im Nahen Osten sprechen.

Target Text:
The two wanted to talk about the implementation of the international agreement and about Teheran's destabilising activities in the Middle East.


We need to preprocess the dataset to prepare it for input into the T5 model. The T5 model expects input in a specific format, including a task prefix. This is because the T5 model is a multi-purpose model, being able to perform several task, so we need to tell it what it must do.

In [10]:
def preprocess_translation(examples: list[dict]) -> dict:
    """Preprocesses a single example for the translation task.

    Args:
        examples list[dict]: A list of dictionaries containing 'de' and 'en' keys with English and German sentences.

    Returns:
        dict[list]: A dictionary of lists with added 'src_texts' and 'tgt_texts' keys for model input and target.

    The function:
    - Adds the task prefix 'translate German to English: ' to the German sentence.
    - Stores the result in 'src_texts'.
    - Copies the English sentence to 'tgt_texts'.
    """
    # Empty dictionary
    texts = {}
    
    # T5 expects a "translate German to English: " prefix
    texts['src_texts'] = ['translate German to English: ' + ex['de'] for ex in examples]
    texts['tgt_texts'] = [ex['en'] for ex in examples]
    return texts

dataset_nmt_preproc = preprocess_translation(dataset_nmt["translation"])


Let's see how we can generate a translation using the t5 model. From a random dataset, we are going to create translations using the different implemented strategies:

In [17]:
# Obtain source and target texts
random_index = random.randint(0, len(dataset_nmt_preproc['src_texts']))
random_src_sentence = dataset_nmt_preproc['src_texts'][random_index]
random_tgt_sentence = dataset_nmt_preproc['tgt_texts'][random_index]

# Obtain the translated sentence
translation_greedy= generate_text(random_src_sentence, strategy='greedy')
translation_beam_search = generate_text(random_src_sentence, strategy='beam', num_beams=5)
translation_temperature = generate_text(random_src_sentence, strategy='temperature', temperature=0.7)
translation_top_k = generate_text(random_src_sentence, strategy='top-k', top_k=50)
translation_top_p = generate_text(random_src_sentence, strategy='top-p', top_p=0.95)
translation_contrastive = generate_text(random_src_sentence, strategy='contrastive', top_k=4, penalty_alpha=0.6)

print(random_src_sentence)
print(f"Original translation: \n{random_tgt_sentence}")
print(f"Greedy Search: \n{translation_greedy}")
print(f"Beam Search: \n{translation_beam_search}")
print(f"Temperature Sampling: \n{translation_temperature}")
print(f"Top-k Sampling: \n{translation_top_k}")
print(f"Top-p Sampling: \n{translation_top_p}")
print(f"Contrastive Search: \n{translation_contrastive}")

translate German to English: Schmidt selbst hatte Lamb in einem von ihm geschriebenen Buch erwähnt, in dem er die "wunderbaren Menschen" erwähnte, mit denen er sein akademisches Leben teilte.
Original translation: 
Schmidt himself had included Lamb in a book he wrote where he acknowledged the "wonderful people" he shared his academic life with.
Greedy Search: 
Schmidt himself had Lamb in a book written by him, in which he mentioned the "wunderful people" who he mentioned with his adolescence.
Beam Search: 
Schmidt himself had Lamb in a book written by him, in which he mentioned the "beautiful people" with whom he was involved.
Temperature Sampling: 
Schmidt himself wrote Lamb in a book written by him, which he would mention and mention 'so beautiful' people, with which he might have participated.
Top-k Sampling: 
Schmidt himself had Lamb mentioned in a 'must meet' book, in which he cited the "beautiful people' with whom his German wife belongs.
Top-p Sampling: 
Schmidt himself wrote La

- Do you see something different between the deterministic and the stochastic strategies? Try different random sentences.
>>> Write your answer here
- *The deterministic methods are: Greedy and Beam Search.*
We can notice that this methods produce more literal translations, especially Greedy. For example, in this last case, we can see that the translation made by Greedy Search is the more literal, without adding any kind of "new" or "invented" information. I think that it is also important to mention that since Greedy always chooses the most probable token, they always produce the same translation. On the other hand, Beam Search explores more continuations, which usually produce more fluent sentences, because this method tries to maximize the end, instead of each step like Greedy. In this case we can see that the sentence, even thought it is still pretty literal, it is more developed, specially in the end where he added "with whom he was involved."
- *The stochastic strategies are: Temperature Sampling, Top-k Sampling, Top-p Sampling and Contrastive Search.*
This methods create more creative or varied sentences, because with stochastic methods we introduce randomness. We can see this for example in the Top-k sampling, where we can see "his German wife belongs", which wasn't mentioned in the original sentence.

### Evaluation Metric: BLEU Score
#### What is BLEU Score?
The BLEU (Bilingual Evaluation Understudy) score is a metric for evaluating the quality of text that has been machine-translated from one language to another. It compares a candidate translation to one or more reference translations and calculates a score based on the overlap of n-grams (contiguous sequences of words).

BLEU-4: Considers up to 4-gram matches between the candidate and reference translations. It provides a balance between precision (matching words) and fluency (maintaining the structure of the language). Using BLEU-4 allows us to capture not just individual word matches (unigrams) but also phrases of up to four words. This makes the evaluation more sensitive to the quality of the translation in terms of both accuracy and fluency.

We'll use the ```sacrebleu``` implementation for a standardized BLEU score calculation (which is BLEU-4). You can check the details [here](https://aclanthology.org/W14-3346.pdf).

In [18]:
bleu_metric = evaluate.load('sacrebleu')

In [19]:
strategies = ['greedy', 'beam', 'temperature', 'top-k', 'top-p', 'contrastive']
results = {}

hyperparameters = {
    'num_beams': 5,
    'temperature': 1.0,
    'top_k': 50,
    'top_p': 0.95,
    'penalty_alpha': 0.6
}

for strategy in strategies:
    print(f"Evaluating strategy: {strategy}")
    predictions = []
    references = []
    for src_text, tgt_text in zip(dataset_nmt_preproc["src_texts"], dataset_nmt_preproc["tgt_texts"]): 
        pred = generate_text(src_text, strategy=strategy, **hyperparameters)
        predictions.append(pred)
        references.append([tgt_text])  # SacreBLEU expects a list of references

    # Compute BLEU-4 score
    bleu = bleu_metric.compute(predictions=predictions, references=references, smooth_method='exp')
    results[strategy] = bleu['score']
    print(f"BLEU-4 score for {strategy}: {bleu['score']:.2f}")


Evaluating strategy: greedy
BLEU-4 score for greedy: 20.99
Evaluating strategy: beam
BLEU-4 score for beam: 21.14
Evaluating strategy: temperature
BLEU-4 score for temperature: 10.52
Evaluating strategy: top-k
BLEU-4 score for top-k: 13.36
Evaluating strategy: top-p
BLEU-4 score for top-p: 9.06
Evaluating strategy: contrastive
BLEU-4 score for contrastive: 19.34


- Seeing the above translations and the BLEU score of the different strategies, which strategy would you choose for this use case?
>>> Write your answer here

If we look at the different BLEU scores, we can see that the strategy with the highest score is Beam Seach, then followed by Greedy and then by Contrastive. This means that deterministic methods have produced higher scores in comparison with the stochastic gradients (except contrastive). 

Higher scores using BLEU indicate more accurate and consistent translations, therefore I would probably choose the beam method, because has a higher BLEU score, and I like the translation that it made in the previous cell.


### Use case 2: Story generation

The ```WritingPrompts``` dataset is a collection of imaginative prompts and corresponding stories from the Reddit community. It contains over 300,000 stories written in response to various prompts, making it suitable for training and evaluating models on creative text generation tasks.

We'll use a subset of the dataset for evaluation purposes.

In [20]:
dataset_st_gen = load_dataset('llm-aes/writing-prompts', split='train[:1%]')

README.md:   0%|          | 0.00/604 [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


(…)-00000-of-00001-72fa4d3bb7c5fd8a.parquet:   0%|          | 0.00/20.9M [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


(…)-00000-of-00001-9975c393c441e11b.parquet:   0%|          | 0.00/88.5k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/232360 [00:00<?, ? examples/s]

Generating sample_length_10_to_292 split:   0%|          | 0/1000 [00:00<?, ? examples/s]

In [21]:
# Randomly select a sample from the dataset
sample = random.choice(dataset_st_gen['prompt'])

print("Sample Text:")
print(sample)

Sample Text:
 My sister found a dead bluejay outside her school . Give it a eulogy .



Now, we have to tell the t5 model to generate text after the input sentence.

In [22]:
def preprocess_story_generation(examples: list[dict]) -> dict:
    """Preprocesses a single example for the story generation task.

    Args:
        examples (list[dict]): A list of dictionaries containing 'prompt' key.

    Returns:
        dict[list]: A dictionary of list with added 'src_texts' for model input.

    The function:
    - Adds the task prefix 'Write a story based on: ' to the prompt.
    """
    # Empty dictionary
    texts = {}
    
    # T5 expects a task prefix
    texts['src_texts'] = ['Write a story based on: ' + ex['prompt'] for ex in examples]
    return texts

dataset_st_gen_preproc = preprocess_story_generation(dataset_st_gen)


As we have no original story to compare how good the model generates a story, you should compare the different decoding strategies by looking at some random stories: 

In [23]:
# Obtain source and target texts
random_index = random.randint(0, len(dataset_st_gen_preproc['src_texts']))
random_src_sentence = dataset_st_gen_preproc['src_texts'][random_index]

# Obtain the translated sentence
story_gen_greedy= generate_text(random_src_sentence, max_length=300, strategy='greedy')
story_gen_beam_search = generate_text(random_src_sentence, max_length=300, strategy='beam', num_beams=5)
story_gen_temperature = generate_text(random_src_sentence, max_length=300, strategy='temperature', temperature=0.7)
story_gen_top_k = generate_text(random_src_sentence, max_length=300, strategy='top-k', top_k=50)
story_gen_top_p = generate_text(random_src_sentence, max_length=300, strategy='top-p', top_p=0.95)
story_gen_contrastive = generate_text(random_src_sentence, max_length=300, strategy='contrastive', top_k=4, penalty_alpha=0.6)

print(random_src_sentence)
print(f"Greedy Search: \n{story_gen_greedy}")
print(f"Beam Search: \n{story_gen_beam_search}")
print(f"Temperature Sampling: \n{story_gen_temperature}")
print(f"Top-k Sampling: \n{story_gen_top_k}")
print(f"Top-p Sampling: \n{story_gen_top_p}")
print(f"Contrastive Search: \n{story_gen_contrastive}")

Write a story based on:  The dashing prince and the evil wizard are both gay and totally in love . Meanwhile , the kidnapped fair maiden is kind of a third wheel ...

Greedy Search: 
The dashing prince and the evil wizard are both gay and totally in love . Meanwhile , the kidnapped fair maiden is kind of a third wheel . Meanwhile , the kidnapped fair maiden is kind of a third wheel . Meanwhile , the kidnapped fair maiden is kind of a third wheel . Meanwhile , the kidnapped fair maiden is kind of a third wheel . Meanwhile , the kidnapped fair maiden is kind of a third wheel . Meanwhile , the kidnapped fair maiden is kind of a third wheel . Meanwhile , the kidnapped fair maiden is kind of a third wheel . Meanwhile , the kidnapped fair maiden is kind of a third wheel . Meanwhile , the kidnapped fair maiden is kind of a third wheel . Meanwhile , the kidnapped fair maiden is kind of a third wheel . Meanwhile , the kidnapped fair maiden is kind of a third wheel . Meanwhile , the kidnapped fa

- Seeing the above generated stories of the different strategies, which strategy would you choose for this use case?
>>> Write your answer here

Seeing the results, I do not like any of these strategies. 
First, if we look at the deterministic methods, we can see that they are constantly repeting the same phrase constantly, like in a loop. This is surprising, because they both performed really well in translation, but they are not really good in generation.

Contrastive search has also produced a loop. It has produced a sentence where they are constantly repeating the last phrase.

However, the methods "temperature sampling", "top-k", and "top-p", do add new parts to the phrase, but they do not make a lot of sense. The three of them add a lot of information that is not related, and some of them produce loops.

If I had to choose one I would probably choose the top-p sampling, because even thought it doesn't make sense and it is not related (like the rest of them), at least it doesn't produce a loop.

### Sensitivity Analysis

Try different hyperparameter values for the decoding strategies, try to optimize the BLEU score for the Neural Machine Translation case and generate better stories in the second use case.

- Which optimal configuration have you found for use case 1? Which are your conclusions based on your analysis?
>>> Write your answer here

- Which optimal configuration have you found for use case 2? Which are your conclusions bsaed on your analysis?
>>> Write your answer here

CASE 1

In [24]:
strategies = ['greedy', 'beam', 'temperature', 'top-k', 'top-p', 'contrastive']
results = {}

hyperparameters = {
    'num_beams': 8,
    'temperature': 1.0,
    'top_k': 50,
    'top_p': 0.95,
    'penalty_alpha': 0.6
}

for strategy in strategies:
    print(f"Evaluating strategy: {strategy}")
    predictions = []
    references = []
    for src_text, tgt_text in zip(dataset_nmt_preproc["src_texts"], dataset_nmt_preproc["tgt_texts"]): 
        pred = generate_text(src_text, strategy=strategy, **hyperparameters)
        predictions.append(pred)
        references.append([tgt_text])  # SacreBLEU expects a list of references

    # Compute BLEU-4 score
    bleu = bleu_metric.compute(predictions=predictions, references=references, smooth_method='exp')
    results[strategy] = bleu['score']
    print(f"BLEU-4 score for {strategy}: {bleu['score']:.2f}")

Evaluating strategy: greedy
BLEU-4 score for greedy: 20.99
Evaluating strategy: beam
BLEU-4 score for beam: 21.86
Evaluating strategy: temperature
BLEU-4 score for temperature: 8.94
Evaluating strategy: top-k
BLEU-4 score for top-k: 14.10
Evaluating strategy: top-p
BLEU-4 score for top-p: 10.55
Evaluating strategy: contrastive
BLEU-4 score for contrastive: 19.34


I'm goingo to compare the obtained results: (Before | After)
- BLEU-4 score for greedy: 20.99 | 20.99
- BLEU-4 score for beam: 21.14 | 21.86
- BLEU-4 score for temperature: 10.52 | 8.94
- BLEU-4 score for top-k: 13.36 | 14.10
- BLEU-4 score for top-p: 9.06 | 10.55
- BLEU-4 score for contrastive: 19.34 | 19.34

We can see that I have managed to improve the methods: beam (my favourite for now), top-k, and top-p. This is very positive, because my favourite in the last tries was the beam, so I'm happy that I have managed to improve its score.

This might be because since I have increased the number of beams, I'm allowing it to explore more, which can be positive for obtaining traductions that are more similar to the reference text.

In [26]:
strategies = ['greedy', 'beam', 'temperature', 'top-k', 'top-p', 'contrastive']
results = {}

hyperparameters = {
    'num_beams': 8,
    'temperature': 1.0,
    'top_k': 30,
    'top_p': 0.80,
    'penalty_alpha': 0.6
}

for strategy in strategies:
    print(f"Evaluating strategy: {strategy}")
    predictions = []
    references = []
    for src_text, tgt_text in zip(dataset_nmt_preproc["src_texts"], dataset_nmt_preproc["tgt_texts"]): 
        pred = generate_text(src_text, strategy=strategy, **hyperparameters)
        predictions.append(pred)
        references.append([tgt_text])  # SacreBLEU expects a list of references

    # Compute BLEU-4 score
    bleu = bleu_metric.compute(predictions=predictions, references=references, smooth_method='exp')
    results[strategy] = bleu['score']
    print(f"BLEU-4 score for {strategy}: {bleu['score']:.2f}")


Evaluating strategy: greedy
BLEU-4 score for greedy: 20.99
Evaluating strategy: beam
BLEU-4 score for beam: 21.86
Evaluating strategy: temperature
BLEU-4 score for temperature: 11.83
Evaluating strategy: top-k
BLEU-4 score for top-k: 10.75
Evaluating strategy: top-p
BLEU-4 score for top-p: 12.48
Evaluating strategy: contrastive
BLEU-4 score for contrastive: 19.87


I'm goingo to compare the obtained results: (Before | After)
- BLEU-4 score for greedy: 20.99 | 20.99
- BLEU-4 score for beam: 21.14 | 21.86
- BLEU-4 score for temperature: 10.52 | 11.83
- BLEU-4 score for top-k: 13.36 | 10.75
- BLEU-4 score for top-p: 9.06 | 12.48
- BLEU-4 score for contrastive: 19.34 | 19.87

We can see that I have managed to improve the methods: beam (my favourite for now), temperature, top-p, and contrastive. This is very positive, because I have managed to improve the score of more models. To achieve this I have decreased the k and the p por top-k and top-p.

This might be because since I have decreased the number of options that they can explore, they are choosing more realiable words, which can make their score increase.

CASE 2

In [25]:
# Obtain source and target texts
random_index = random.randint(0, len(dataset_st_gen_preproc['src_texts']))
random_src_sentence = dataset_st_gen_preproc['src_texts'][random_index]

# Obtain the translated sentence
story_gen_greedy= generate_text(random_src_sentence, max_length=300, strategy='greedy')
story_gen_beam_search = generate_text(random_src_sentence, max_length=300, strategy='beam', num_beams=5)
story_gen_temperature = generate_text(random_src_sentence, max_length=300, strategy='temperature', temperature=0.4) # I HAVE CHANGED THIS TO A LOWER VALUES, BECAUSE WITH A LOWER VALUE IT SHOULD PRODUCE A MORE COHERENT TEXT.
story_gen_top_k = generate_text(random_src_sentence, max_length=300, strategy='top-k', top_k=30) # I HAVE CHANGED THIS TO A LOWER VALUE, BECAUSE MAYBE LIKE THAT IT COULD A LESS RANDOM PHRASE, BUT I HAVE TO BE CAREFUL SO THAT IT DOESN'T INCREASE THE NUMBER OF LOOPS
story_gen_top_p = generate_text(random_src_sentence, max_length=300, strategy='top-p', top_p=0.8) # I HAVE CHANGED THIS TO A LOWER VALUE SO THAT IT IS MORE CONTROLED.
story_gen_contrastive = generate_text(random_src_sentence, max_length=300, strategy='contrastive', top_k=5, penalty_alpha=0.7) # I HAVE INCREASED BOTH VALUES, SO THAT IT EXPLORES MORE, BUT IT DOESN'T PRODUCE LOOPS.

print(random_src_sentence)
print(f"Greedy Search: \n{story_gen_greedy}")
print(f"Beam Search: \n{story_gen_beam_search}")
print(f"Temperature Sampling: \n{story_gen_temperature}")
print(f"Top-k Sampling: \n{story_gen_top_k}")
print(f"Top-p Sampling: \n{story_gen_top_p}")
print(f"Contrastive Search: \n{story_gen_contrastive}")

Write a story based on:  Write the opening scene to the yet-unwritten novel you 've been thinking about . Introduce us to your main character , and show us what kind of book it 's going to be .

Greedy Search: 
a novel is a novel that 's a bit of a mystery . It 's a novel that 's a bit of a mystery . It 's a mystery that 's a bit of a mystery . It 's a mystery that 's a bit of a mystery . It 's a mystery that 's a bit of a mystery .
Beam Search: 
This is the opening scene of the novel you 've been thinking about . It 's about a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a young woman , a yo

We can see that the greedy search and the beam search make loops and repetitions. This is something that usually happens when the model is too deterministic, because we are always choosing the most probable token, which could be positive for translation, but not positive for creative generation.

For temperature sampling, the output it's a little bit more coherent because it is trying to explain what the book is about, but it is also very repetitive, because it creates a loop.

For the top-k, it creates more structure and variation, because it tries to explain what the story is about and uses a big variety of adjectives. After that it starts saying things without sense, but at least is not repetitive.

Top-p sampling is not working very well, because it tells you that it is going to show what is going to be about, but it stops there.

Lastly, the contrastive search. It does tell what the story is about, and presents a little bit the character, but then it does a loop.
